com.unity.perception/Runtime/Randomization/Parameters/TypedParameter.cs

```diff
@@ -0,0 +1,37 @@
+﻿using System;
+using UnityEngine.Perception.Randomization.Samplers;
+
+namespace UnityEngine.Perception.Randomization.Parameters
+{
+    public abstract class TypedParameter<T> : Parameter
+    {
+        public sealed override Type OutputType => typeof(T);
+
+        /// <summary>
+        /// Generates one parameter sample
+        /// </summary>
+        /// <param name="index">Often the current scenario iteration or a scenario's framesSinceInitialization</param>
+        public abstract T Sample(int index);
+
+        /// <summary>
+        /// Generates an array of parameter samples
+        /// </summary>
+        /// <param name="index">Often the current scenario iteration or a scenario's framesSinceInitialization</param>
+        /// <param name="sampleCount">Number of parameter samples to generate</param>
+        public abstract T[] Samples(int index, int sampleCount);
```

Comment:
You can't return a list of managed samples (like materials) with a NativeArray.

Reply:
I did a performance test comparing the generation of GC arrays vs returning an Enumerable object and iterating over each sample. I found that creating and returning the array was faster than generating samples while iterating over the enumerable.

Reply:
I recently updated parameters to have a stateful Sample() method, meaning users can just for loop and call Sample() repeatedly instead of generating samples up front. This method is not longer necessary and has been removed.